### PR TITLE
New algorithm to find face-cell adjacencies

### PR DESCRIPTION
**Pull request by OrigamiDraw**
Function X.V_FV_P_CP_2_CF_FC(Vf, FV, P, CP) is replaced with a new function X.EF_FV_SP_SE_CP_SC_2_CF_FC(EF, FV, SP, SE, CP, SC).
Function M.convex_contains_point(P, q), now unused, is removed.
This addresses Issue #10 but with a different algorithm.

The new function finds the face-cell adjacencies using the following algorithm:

• We use an SC_map that lists all Segments of the overlap graph, along with the one or two Cells they delimit
• We use an SF_map that lists all Segments of the overlap graph, along with all the Faces they delimit. We obtain this by using SE, which gives the correspondence between the Segments in the overlap graph and the Edges in the CP
• We pick a Segment that has only one Cell associated in SC_map, i.e., that is on the border of the overlap graph
• We start in that Cell and find all Faces that overlap it by using SF_map for the chosen Segment
• BFS:
    o Pick another interior Segment of the current Cell that hasn't been visited yet
    o When we cross that Segment, we know which Cell we enter by SC_map
    o We also know all the Faces that end on this Segment by SF_map. Depending on whether we were inside or outside these faces, we know if we enter or exit them. So, we can build CF for all the Faces we enter
    o When all interior Segments of the current Cell have been visited, we continue with one of the Cells adjacent to it
    o We continue until we visit all the Cells
• We sort CF
• We build FC from CF

### CODE REVIEW

**Comment by origamimagiro** [comment]
Looks great! This is an excellent implementation! The use of `SF_map` is particularly elegant, I agree! :)
I have no notes on the algorithm, but I have a few comments on code style: 

- **Request:** I would prefer we avoid the use of [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), e.g. line `810`. In many browsers `Array.push(...A)` is implemented recursively, so for `A` with large length, it can grow the call stack past the allowable limit, which I'd like to avoid. It's not a problem here, since `EF[ei].length` can be at most `2`, but if you don't mind, I'd prefer to avoid spread syntax completely in this code base (i.e., replace this instance with a `for` loop over the elements of `EF[ei]`). 
- **Comment:** The construction of `SC_map` is essentially identical to `EF_map` in `V_VF_EV_EA_2_Vf_Ff` (though I will note that your implementation is actually more consistent with the rest of the code base). We may want to refactor this into a separate method we can call twice, but we don't have to do that with this change. 
- **Request:** `const CF = CP.map((c) => undefined);` when arguments are not used, this code base leaves them off, so `const CF = CP.map(() => undefined);`. Same for line `856` in the declaration of `FC`.
- **Request:** Line `846` is getting pretty long. Consider `const k = M.encode_order_pair([v1, v2])` and `for (const f of SF_map.get(k)) {`
- **Request:** `if (!Fs.delete(f)) {` in line `846` is a little hard to parse. Usually this codebase would do the more verbose:
   ```
   if (Fs.has(f)) {
       Fs.delete(f);
   } else {
       Fs.add(f);
   }
   ```
   I know this is technically slower because it may lookup `f` twice, but it is much more readable. 

If you agree with these changes, I will proceed with the merge. You can either implement the changes in your pull request, or I can make them after the merge.

**Comment by OrigamiDraw** [comment]
OK, I understand these requests. Please go ahead; I am not sure how to modify the pull request.

**Comment by origamimagiro** [comment]
Done!

**Comment by OrigamiDraw** [comment]
Thanks. What is your maximum line length?

**Comment by origamimagiro** [comment]
80 characters, which seems to be at least one [standard width](https://softwareengineering.stackexchange.com/questions/148677/why-is-80-characters-the-standard-limit-for-code-width). I can't remember why I adopted this standard, but now I'm used to it...  ¯\_(ツ)_/¯
